hack/generator/pkg/astmodel/type_visitor.go

```diff
@@ -188,17 +211,21 @@ func IdentityVisitOfResourceType(this *TypeVisitor, it *ResourceType, ctx interf
 }
 
 func IdentityVisitOfArmType(this *TypeVisitor, at *ArmType, ctx interface{}) (Type, error) {
-	newType, err := this.Visit(&at.objectType, ctx)
+	nt, err := this.Visit(&at.objectType, ctx)
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to visit ARM underlying type %v", at.objectType)
 	}
 
-	ot, ok := newType.(*ObjectType)
-	if !ok {
-		return nil, errors.Errorf("expected transformation of ARM underlying type %v to return ObjectType, not %v", at.objectType, ot)
-	}
+	switch newType := nt.(type) {
+	case *ObjectType:
+		return MakeArmType(*newType), nil
+
+	case *ArmType:
```

Comment:
I added this in because it was needed - because I had the default case (below) return an error. 

More generally, the signature of the methods for `TypeVisitor` allow any transformation, so it's perfectly valid for an `ObjectType` to go in, and an `ArmType` to come out (and in fact we do this in a few different places).

Reply:
Note that we have transformations that take in an ObjectType and return an ArmType (part of the pipeline to create the ArmTypes); the same is happening with StorageTypes as well.